plugins/crypto/src/main/java/org/opensearch/cryptoplugin/CryptoBasePlugin.java

```diff
@@ -0,0 +1,168 @@
+import org.opensearch.threadpool.FixedExecutorBuilder;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.watcher.ResourceWatcherService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * AWS plugin for encryption and decryption support.
+ */
+public class CryptoBasePlugin extends Plugin implements CryptoPlugin, ExtensiblePlugin {
+    private static final Logger logger = LogManager.getLogger(CryptoBasePlugin.class);
+    private static final String NAME = "crypto";
+    private final Map<String, CryptoKeyProviderExtension> keyProviderExtensions;
+    protected final Map<String, Map<String, CryptoManagerImpl>> keyProviderCryptoManagers;
+    private final long cacheTTL = TimeUnit.DAYS.toMillis(7);
+    private CacheRefresher cacheRefresher;
+    private static final String CRYPTO_ASYNC_REFRESH_THREADPOOL = "crypto_async_refresh";
+
+    /**
+     * Constructor for crypto
+     */
+    public CryptoBasePlugin() {
+        keyProviderExtensions = new HashMap<>();
+        keyProviderCryptoManagers = new HashMap<>();
+    }
+
+    @Override
+    public List<ExecutorBuilder<?>> getExecutorBuilders(Settings settings) {
+        List<ExecutorBuilder<?>> executorBuilders = new ArrayList<>();
+        executorBuilders.add(new FixedExecutorBuilder(settings, CRYPTO_ASYNC_REFRESH_THREADPOOL, 1, 10, CRYPTO_ASYNC_REFRESH_THREADPOOL));
+        return executorBuilders;
+    }
+
+    @Override
+    public Collection<Object> createComponents(
+        final Client client,
+        final ClusterService clusterService,
+        final ThreadPool threadPool,
+        final ResourceWatcherService resourceWatcherService,
+        final ScriptService scriptService,
+        final NamedXContentRegistry xContentRegistry,
+        final Environment environment,
+        final NodeEnvironment nodeEnvironment,
+        final NamedWriteableRegistry namedWriteableRegistry,
+        final IndexNameExpressionResolver expressionResolver,
+        final Supplier<RepositoriesService> repositoriesServiceSupplier
+    ) {
+        cacheRefresher = new CacheRefresher(CRYPTO_ASYNC_REFRESH_THREADPOOL, threadPool);
+        return Collections.emptyList();
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public CryptoManager.Factory createClientFactory(String keyProviderType) {
+        return (cryptoSettings, keyProviderName) -> {
+            if (!keyProviderExtensions.containsKey(keyProviderType)) {
+                throw new IllegalArgumentException(
+                    "Key provider extension of type " + keyProviderType + " not found in the installed crypto extension plugins."
+                );
+            }
+
+            synchronized (keyProviderCryptoManagers) {
```

Comment:
Curious where is the instantiation being done? Is it in a multi-threaded context?

Reply:
It is being done during repo registration. Yes, it is in a multi-threaded context as there can be multiple simultaneous  repo registration events.